### PR TITLE
JDK-8301444: Port fdlibm hyperbolic transcendental functions to Java

### PR DESCRIPTION
**Pull request by jddarcy**
Initial pass of porting FDLIBM sinh/cosh/tanh to Java. I do intend to refactor the regression tests a bit to reduce duplication, but the actual ports should be ready for review.

Diff'ing the ports as before, original vs transliteration port:

```
$ diff -w Hyperbolic.c Hyperbolic.translit.java
1c1
< /* __ieee754_sinh(x)
---
>     /**
17a18,19
>     static class Sinh {
>         private static final double one = 1.0, shuge = 1.0e307;
19,33c21
< #include "fdlibm.h"
< 
< #ifdef __STDC__
< static const double one = 1.0, shuge = 1.0e307;
< #else
< static double one = 1.0, shuge = 1.0e307;
< #endif
< 
< #ifdef __STDC__
<         double __ieee754_sinh(double x)
< #else
<         double __ieee754_sinh(x)
<         double x;
< #endif
< {
---
>         private static double compute(double x) {
36c24
<         unsigned lx;
---
>             /* unsigned */ int lx;
51c39
<             t = expm1(fabs(x));
---
>                 t = FdlibmTranslit.expm1(Math.abs(x));
57c45
<         if (ix < 0x40862E42)  return h*__ieee754_exp(fabs(x));
---
>             if (ix < 0x40862E42)  return h*StrictMath.exp(Math.abs(x)); // TODO switch to translit
60,62c48,52
<         lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
<         if (ix<0x408633CE || ((ix==0x408633ce)&&(lx<=(unsigned)0x8fb9f87d))) {
<             w = __ieee754_exp(0.5*fabs(x));
---
>             // lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
>             // lx =  (((*(unsigned*)&one)>>29)) + (unsigned*)&x ;
>             lx = __LO(x);
>             if (ix<0x408633CE || ((ix==0x408633ce)&&(Long.compareUnsigned(lx, 0x8fb9f87d) <= 0 ))) {
>                 w = StrictMath.exp(0.5*Math.abs(x)); // TODO switch to translit
70c60,62
< /* __ieee754_cosh(x)
---
>     }
> 
>     /**
90,105c82,84
< 
< #include "fdlibm.h"
< 
< #ifdef __STDC__
< static const double one = 1.0, half=0.5, huge = 1.0e300;
< #else
< static double one = 1.0, half=0.5, huge = 1.0e300;
< #endif
< 
< #ifdef __STDC__
<         double __ieee754_cosh(double x)
< #else
<         double __ieee754_cosh(x)
<         double x;
< #endif
< {
---
>     static class Cosh {
>         private static final double one = 1.0, half=0.5, huge = 1.0e300;
>         private static double compute(double x) {
108c87
<         unsigned lx;
---
>             /*unsigned*/ int lx;
119c98
<             t = expm1(fabs(x));
---
>                 t = expm1(Math.abs(x));
127c106
<                 t = __ieee754_exp(fabs(x));
---
>                 t = StrictMath.exp(Math.abs(x)); // TODO switch to translit
132c111
<         if (ix < 0x40862E42)  return half*__ieee754_exp(fabs(x));
---
>             if (ix < 0x40862E42)  return half*StrictMath.exp(Math.abs(x)); // TODO switch to translit
135c114
<         lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
---
>             lx = __LO(x);
137,138c116,117
<               ((ix==0x408633ce)&&(lx<=(unsigned)0x8fb9f87d))) {
<             w = __ieee754_exp(half*fabs(x));
---
>                 ((ix==0x408633ce)&&(Integer.compareUnsigned(lx, 0x8fb9f87d) <= 0))) {
>                 w = StrictMath.exp(half*Math.abs(x)); // TODO switch to translit
146c125,127
< /* Tanh(x)
---
>     }
> 
>     /**
169,184c150,152
< 
< #include "fdlibm.h"
< 
< #ifdef __STDC__
< static const double one=1.0, two=2.0, tiny = 1.0e-300;
< #else
< static double one=1.0, two=2.0, tiny = 1.0e-300;
< #endif
< 
< #ifdef __STDC__
<         double tanh(double x)
< #else
<         double tanh(x)
<         double x;
< #endif
< {
---
>     static class Tanh {
>         private static final double one=1.0, two=2.0, tiny = 1.0e-300;
>         static double compute(double x) {
203c171
<                 t = expm1(two*fabs(x));
---
>                     t = expm1(two*Math.abs(x));
206c174
<                 t = expm1(-two*fabs(x));
---
>                     t = expm1(-two*Math.abs(x));
214a183
>     }
```

Note that the original has a in-line version of the "__LO" macro rather than using the macro.

And transliteration vs more idiomatic:

```
$ diff -w Hyperbolic.translit.java Hyperbolic.fdlibm.java 
21c21
<         private static double compute(double x) {
---
>          static double compute(double x) {
26c26
<             /* High word of |x|. */
---
>             // High word of |x|
28c28
<             ix = jx&0x7fffffff;
---
>             ix = jx & 0x7fff_ffff;
30,31c30,33
<             /* x is INF or NaN */
<             if(ix>=0x7ff00000) return x+x;
---
>             // x is INF or NaN
>             if ( ix >= 0x7ff0_0000) {
>                 return x + x;
>             }
34,40c36,48
<             if (jx<0) h = -h;
<             /* |x| in [0,22], return sign(x)*0.5*(E+E/(E+1))) */
<             if (ix < 0x40360000) {          /* |x|<22 */
<                 if (ix<0x3e300000)          /* |x|<2**-28 */
<                     if(shuge+x>one) return x;/* sinh(tiny) = tiny with inexact */
<                 t = FdlibmTranslit.expm1(Math.abs(x));
<                 if(ix<0x3ff00000) return h*(2.0*t-t*t/(t+one));
---
>             if ( jx < 0) {
>                 h = -h;
>             }
>             // |x| in [0,22], return sign(x)*0.5*(E+E/(E+1)))
>             if (ix < 0x4036_0000) {          // |x| < 22
>                 if (ix < 0x3e30_0000)        // |x| < 2**-28
>                     if (shuge + x > one) {   // sinh(tiny) = tiny with inexact
>                         return x;
>                     }
>                 t = StrictMath.expm1(Math.abs(x));
>                 if (ix < 0x3ff0_0000) {
>                     return h*(2.0 * t - t*t/(t + one));
>                 }
44,45c52,55
<             /* |x| in [22, log(maxdouble)] return 0.5*exp(|x|) */
<             if (ix < 0x40862E42)  return h*StrictMath.exp(Math.abs(x)); // TODO switch to translit
---
>             // |x| in [22, log(maxdouble)] return 0.5*exp(|x|)
>             if (ix < 0x4086_2E42) {
>                 return h*StrictMath.exp(Math.abs(x));
>             }
47,49c57
<             /* |x| in [log(maxdouble), overflowthresold] */
<             // lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
<             // lx =  (((*(unsigned*)&one)>>29)) + (unsigned*)&x ;
---
>             // |x| in [log(maxdouble), overflowthresold]
51,52c59,62
<             if (ix<0x408633CE || ((ix==0x408633ce)&&(Long.compareUnsigned(lx, 0x8fb9f87d) <= 0 ))) {
<                 w = StrictMath.exp(0.5*Math.abs(x)); // TODO switch to translit
---
>             if (ix < 0x4086_33CE ||
>                 ((ix == 0x4086_33ce) &&
>                  (Long.compareUnsigned(lx, 0x8fb9_f87d) <= 0 ))) {
>                 w = StrictMath.exp(0.5 * Math.abs(x));
57c67
<             /* |x| > overflowthresold, sinh(x) overflow */
---
>             // |x| > overflowthresold, sinh(x) overflow
84c94
<         private static double compute(double x) {
---
>         static double compute(double x) {
89c99
<             /* High word of |x|. */
---
>             // High word of |x|
91c101
<             ix &= 0x7fffffff;
---
>             ix &= 0x7fff_ffff;
93,94c103,106
<             /* x is INF or NaN */
<             if(ix>=0x7ff00000) return x*x;
---
>             // x is INF or NaN
>             if (ix >= 0x7ff_00000) {
>                 return x*x;
>             }
96,98c108,110
<             /* |x| in [0,0.5*ln2], return 1+expm1(|x|)^2/(2*exp(|x|)) */
<             if(ix<0x3fd62e43) {
<                 t = expm1(Math.abs(x));
---
>             // |x| in [0,0.5*ln2], return 1+expm1(|x|)^2/(2*exp(|x|))
>             if (ix < 0x3fd6_2e43) {
>                 t = StrictMath.expm1(Math.abs(x));
100c112,114
<                 if (ix<0x3c800000) return w;        /* cosh(tiny) = 1 */
---
>                 if (ix < 0x3c80_0000) { // cosh(tiny) = 1
>                     return w;
>                 }
104,106c118,120
<             /* |x| in [0.5*ln2,22], return (exp(|x|)+1/exp(|x|)/2; */
<             if (ix < 0x40360000) {
<                 t = StrictMath.exp(Math.abs(x)); // TODO switch to translit
---
>             // |x| in [0.5*ln2, 22], return (exp(|x|) + 1/exp(|x|)/2
>             if (ix < 0x4036_0000) {
>                 t = StrictMath.exp(Math.abs(x));
110,111c124,127
<             /* |x| in [22, log(maxdouble)] return half*exp(|x|) */
<             if (ix < 0x40862E42)  return half*StrictMath.exp(Math.abs(x)); // TODO switch to translit
---
>             // |x| in [22, log(maxdouble)] return half*exp(|x|)
>             if (ix < 0x4086_2E42) {
>                 return half*StrictMath.exp(Math.abs(x));
>             }
113c129
<             /* |x| in [log(maxdouble), overflowthresold] */
---
>             // |x| in [log(maxdouble), overflowthresold]
115,117c131,134
<             if (ix<0x408633CE ||
<                 ((ix==0x408633ce)&&(Integer.compareUnsigned(lx, 0x8fb9f87d) <= 0))) {
<                 w = StrictMath.exp(half*Math.abs(x)); // TODO switch to translit
---
>             if (ix<0x4086_33CE ||
>                 ((ix == 0x4086_33ce) &&
>                  (Integer.compareUnsigned(lx, 0x8fb9_f87d) <= 0))) {
>                 w = StrictMath.exp(half*Math.abs(x));
122c139
<             /* |x| > overflowthresold, cosh(x) overflow */
---
>             // |x| > overflowthresold, cosh(x) overflow
126d142
< 
156c172
<             /* High word of |x|. */
---
>             // High word of |x|.
158c174
<             ix = jx&0x7fffffff;
---
>             ix = jx & 0x7fff_ffff;
160,163c176,182
<             /* x is INF or NaN */
<             if(ix>=0x7ff00000) {
<                 if (jx>=0) return one/x+one;    /* tanh(+-inf)=+-1 */
<                 else       return one/x-one;    /* tanh(NaN) = NaN */
---
>             // x is INF or NaN
>             if (ix >= 0x7ff0_0000) {
>                 if (jx >= 0) {  // tanh(+-inf)=+-1
>                     return one/x + one;
>                 } else {        // tanh(NaN) = NaN
>                     return one/x - one;
>                 }
166,171c185,190
<             /* |x| < 22 */
<             if (ix < 0x40360000) {          /* |x|<22 */
<                 if (ix<0x3c800000)          /* |x|<2**-55 */
<                     return x*(one+x);       /* tanh(small) = small */
<                 if (ix>=0x3ff00000) {       /* |x|>=1  */
<                     t = expm1(two*Math.abs(x));
---
>             // |x| < 22
>             if (ix < 0x4036_0000) {          // |x| < 22
>                 if (ix<0x3c80_0000)          // |x| < 2**-55
>                     return x*(one + x);      // tanh(small) = small
>                 if (ix>=0x3ff0_0000) {       // |x| >= 1
>                     t = StrictMath.expm1(two*Math.abs(x));
174c193
<                     t = expm1(-two*Math.abs(x));
---
>                     t = StrictMath.expm1(-two*Math.abs(x));
177,179c196,197
<                 /* |x| > 22, return +-1 */
<             } else {
<                 z = one - tiny;             /* raised inexact flag */
---
>             } else { // |x| > 22, return +-1
>                 z = one - tiny;             // raised inexact flag
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301444](https://bugs.openjdk.org/browse/JDK-8301444): Port fdlibm hyperbolic transcendental functions to Java

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12429/head:pull/12429` \
`$ git checkout pull/12429`

Update a local copy of the PR: \
`$ git checkout pull/12429` \
`$ git pull https://git.openjdk.org/jdk pull/12429/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12429`

View PR using the GUI difftool: \
`$ git pr show -t 12429`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12429.diff">https://git.openjdk.org/jdk/pull/12429.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/lang/StrictMath/HyperbolicTests.java [hunk 1]
```diff
@@ -42,6 +50,243 @@
...
+    private static int testAgainstTranslitSinh() {
+        int failures = 0;
+        double x;
+
+        // Test just above subnormal threshold...
+        x = Double.MIN_NORMAL;
+        failures += testRangeSinh(x, Math.ulp(x), 1000);
+
+         // ... and just below subnormal threshold ...
+          x =  Math.nextDown(Double.MIN_NORMAL);
```
Comment (turbanoff) [inline test/jdk/java/lang/StrictMath/HyperbolicTests.java:86]:
```suggestion
          x = Math.nextDown(Double.MIN_NORMAL);
```

File: test/jdk/java/lang/StrictMath/HyperbolicTests.java [hunk 2]
```diff
@@ -42,6 +50,243 @@
...
+    private static int testAgainstTranslitCosh() {
+        int failures = 0;
+        double x;
+
+        // Test just above subnormal threshold...
+        x = Double.MIN_NORMAL;
+        failures += testRangeCosh(x, Math.ulp(x), 1000);
+
+         // ... and just below subnormal threshold ...
+          x =  Math.nextDown(Double.MIN_NORMAL);
```
Comment (turbanoff) [inline test/jdk/java/lang/StrictMath/HyperbolicTests.java:140]:
```suggestion
          x = Math.nextDown(Double.MIN_NORMAL);
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 3]
```diff
@@ -1194,4 +1194,206 @@ static double compute(double x) {
...
+            jx = __HI(x);
+            ix = jx & 0x7fff_ffff;
+
+            // x is INF or NaN
+            if ( ix >= 0x7ff0_0000) {
+                return x + x;
+            }
+
+            h = 0.5;
+            if ( jx < 0) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:1233]:
```suggestion
            if (jx < 0) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 4]
```diff
@@ -1194,4 +1194,206 @@ static double compute(double x) {
...
+            double t, w, h;
+            int ix, jx;
+            /* unsigned */ int lx;
+
+            // High word of |x|
+            jx = __HI(x);
+            ix = jx & 0x7fff_ffff;
+
+            // x is INF or NaN
+            if ( ix >= 0x7ff0_0000) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:1228]:
```suggestion
            if (ix >= 0x7ff0_0000) {
```

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java [hunk 5]
```diff
@@ -808,4 +821,188 @@ else if(hx < 0x3c900000) {      /* when |x|<2**-54, return x */
...
+                return h*(t+t/(t+one));
+            }
+
+            /* |x| in [22, log(maxdouble)] return 0.5*exp(|x|) */
+            if (ix < 0x40862E42)  return h*StrictMath.exp(Math.abs(x)); // TODO switch to translit
+
+            /* |x| in [log(maxdouble), overflowthresold] */
+            // lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
+            // lx =  (((*(unsigned*)&one)>>29)) + (unsigned*)&x ;
+            lx = __LO(x);
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/FdlibmTranslit.java:874]: This aspect of the transliteration is unclear.
The C code seems to ignore endianness of `double` values, and there's a shift by 29 that seems intentional for some reason.
Are we safe with the `__LO(x)` transliteration?

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java [hunk 6]
```diff
@@ -808,4 +821,188 @@ else if(hx < 0x3c900000) {      /* when |x|<2**-54, return x */
...
+            if (ix < 0x40360000) {
+                t = StrictMath.exp(Math.abs(x)); // TODO switch to translit
+                return half*t+half/t;
+            }
+
+            /* |x| in [22, log(maxdouble)] return half*exp(|x|) */
+            if (ix < 0x40862E42)  return half*StrictMath.exp(Math.abs(x)); // TODO switch to translit
+
+            /* |x| in [log(maxdouble), overflowthresold] */
+            lx = __LO(x);
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/FdlibmTranslit.java:938]: Same concern as above about the transliteration to `__LO(x)`

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 7]
```diff
@@ -1194,4 +1194,206 @@ static double compute(double x) {
...
+            double t, w;
+            int ix;
+            /*unsigned*/ int lx;
+
+            // High word of |x|
+            ix = __HI(x);
+            ix &= 0x7fff_ffff;
+
+            // x is INF or NaN
+            if (ix >= 0x7ff_00000) {
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/FdLibm.java:1301]: For consistency with L.1228 and L.1374:
`ix >= 0x7ff0_0000`

File: src/java.base/share/classes/java/lang/FdLibm.java (hunk 3 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/FdLibm.java]: Spacing fixed; thanks @turbanoff .

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java (hunk 5 above)
Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/FdlibmTranslit.java:874]: Yes, it gave me pause when doing the transliteration.

Unpacking the code a bit,

```
            /* |x| in [log(maxdouble), overflowthresold] */
            // lx = *( (((*(unsigned*)&one)>>29)) + (unsigned*)&x);
            lx = __LO(x);
            if (ix<0x408633CE || 
            ((ix==0x408633ce)&&(Long.compareUnsigned(lx, 0x8fb9f87d) <= 0 ))) {...}
```

Assuming the comment accurate describe the intention of the code, identifying values between log(maxdouble) and the overflow threshold. Those particular values are as decimal fp, hex hp, and long bits:

709.782712893384    0x1.62e42fefa39efp9    40862e42__fefa39ef
710.4758600739439    0x1.633ce8fb9f87dp9    408633ce_8fb9f87d

so the conditional is checking if the high word (ix) is for something less than the low end of the range OR if the high word matches the high word for the high end of the range AND low bits are for less than the high end of the range. Therefore, I think taking __LO(x) is the correct semantics here.

FWIW, the ExhaustingTest of running all the float values against sinh/cosh/tanh in the transliteration port passes when using JDK 20 baseline as a reference.

Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/FdlibmTranslit.java:874]: PS One possible future refactoring to the code in

src/java.base/share/classes/java/lang/FdLibm.java

would be to replace such integer-based range checks with floating-point based range checks, in this case something like:

```
if (absX <= 0x1.633ce8fb9f87dp9) { // 710.4758600739439
    ...}
```

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java (hunk 6 above)
Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/FdlibmTranslit.java:938]: Right; sinh and cosh overflow at the same input.

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java (hunk 5 above)
Comment (rgiulietti) [reply test/jdk/java/lang/StrictMath/FdlibmTranslit.java:874]: Let's discuss this C fragment.
`*( (((*(unsigned*)&one)>>29)) + (unsigned*)&x)`

On a big-endian platform, `*(unsigned*)&one` returns `__HI(one) = 0x3ff0_0000`. It is then shifted right by 29, giving `1`. And `(unsigned*)&x` is the _address_ of the high word of `x`. The sum with `1` thus gives the address of the low word of `x`, which is dereferenced by the leftmost `*`, resulting in `__LO(x)`.

On a little-endian platform, `*(unsigned*)&one` returns `__LO(one) = 0x0`. It is then shifted right by 29, giving `0`. And `(unsigned*)&x` is the _address_ of the low word of `x`. The sum with `0` thus gives the address of the low word of `x`, which is dereferenced by the leftmost `*`, resulting in `__LO(x)` as well.

In other words, this is a complicated way to say `__LO(x)`, right? I wonder why the author came up with this creative code...

Would it make sense to add this discussion as a block comment?

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 8]
```diff
@@ -1194,4 +1194,206 @@ static double compute(double x) {
...
+     *
+     *          22       <= x <= lnovft :  sinh(x) := exp(x)/2
+     *          lnovft   <= x <= ln2ovft:  sinh(x) := exp(x/2)/2 * exp(x/2)
+     *          ln2ovft  <  x           :  sinh(x) := x*shuge (overflow)
+     *
+     * Special cases:
+     *      sinh(x) is |x| if x is +INF, -INF, or NaN.
+     *      only sinh(0)=0 is exact for finite x.
+     */
+    static class Sinh {
```
Comment (magicus) [inline src/java.base/share/classes/java/lang/FdLibm.java:1215]: For clarity, should this not be declared `final`?

Comment (jddarcy) [reply src/java.base/share/classes/java/lang/FdLibm.java:1215]: Hmm. Wouldn't hurt to do so.

I was thinking of making a pass over Fdlibm.java at some point to make sure the nested classes had package-level rather than public accessibility and that the classes had private constructors that threw an exception (to prevent instantiation).

Comment (magicus) [reply src/java.base/share/classes/java/lang/FdLibm.java:1215]: It doesn't matter much. It was just a drive-by comment; I don't normally look at this part of the code but the PR title attracted my attention (it is not everyday you see hyperbolic transcendental PRs in the JDK), and when I skimmed through the code, some reflex part of my brain said "utility classes should be final".

Comment (jddarcy) [reply src/java.base/share/classes/java/lang/FdLibm.java]: Changed the nested classes for the hyperbolic functions to be final and added a private constructor.

**Review by bplb — APPROVED** [review]
Approved.
